docs/source/design/user_messaging.rst

```diff
@@ -0,0 +1,124 @@
+.. -*- mode: rst -*-
+.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:
+
+.. _chap_design_user_messaging:
+
+*******************************************************
+User messaging: result records vs exceptions vs logging
+*******************************************************
+
+.. topic:: Specification scope and status
+
+   This specification aims to delineate the applicable contexts for using result records,
+   exceptions, logging, or other types of user messaging methods.
+
+Motivation
+==========
+
+Design specifications exist for :ref:`chap_design_result_records`,
+:ref:`chap_design_exception_handling`, :ref:`chap_design_log_levels`, and 
+:ref:`chap_design_progress_reporting`, and an ideal yet elusive goal is the consistent
+application of these methods in the applicable contexts. 
+
+From a **developer's perspective**, a specification for the proper use of specific user
+messaging methods creates a best practice standard that simplifies the implementation
+process, minimizes code duplication, and ensures consistency in the code base. This
+applies both to development in DataLad core as well as in DataLad extensions.
+
+From a **user's perspective**, it is imperative to receive the most appropriate and
+unambiguous message relating to the current context or action.
+
+Discrepancies (on the development side) and resulting ambiguity (on the user side) arise
+when the chosen user messaging methods are implemented interchangeably, inconsistently,
+within contexts that do not warrant them, and in ways that do not inform users
+appropriately.
+
+Consequently, the motivation of this specification is to delineate the applicable
+contexts for using result records, exceptions, logging, or other types of
+user messaging processes.
+
+
+Specification
+=============
+
+Result records
+--------------
+
+**Result records are the standard and preferred return value** format for all
+DataLad commands. Result records are routinely inspected throughout the code base,
+and are used to inform generic error handling, as well as particular calling commands
+on how to proceed with a specific operation.
+
+Yield result records when:
+
+1. A generic and standard way of error handling is considered useful
+2. A process can reasonably be continued despite evident errors or shortcomings in
```

Comment:
maybe this warrants further explanation? E.g., mention that "on_failure=ignore" or "result_renderer=disabled" provide means to purposefully ignore failures

Reply:
moreover even if it can't continue, it would then just yield and return, i.e. still not resort to an exception. I.e.  not necessarily it can continue but still might need to yield record instead of just raising an exception.

But then taken together - it remains unclear when to raise Exception vs when yield record(s). 

I feel that there is more of us having two notions of functions (or "logic flow") in the codebase:
- *classic Python interfaces* -- return actual values, expected to interrupt operation and raise an `Exception` in case of error, exception bubbles up in the traceback automatically until explicitly handled so logic flow is bubbling up too.
- *datalad interfaces* -- message passing by yielding result records associated with individual items, could potentially continue operation after an error (on one of the items, or just to the next steps in logic).

And then it is pretty much for a developer to choose classical vs the datalad-interface style depending 
- how close to `Interface` layer/code it resides
- that 2nd factor of needing to be able to continue operation in case of an error relating to a specific item, so pretty much for any function where there is a loop over items, OR
- usefulness to preparing full "message" with not only a value but some possible extra annotation (hints etc) which otherwise would not be assignable to a basic return type in the *classical* one.

So I think it is valuable to describe such separation of interfaces, and clarification that *datalad interfaces* is pretty much "message passing" flow of logic, while original *classic* logic results in non-annotated return values, and interrupted logic in case of individual item(s) erroring out.

Reply:
I like this suggestion of describing the datalad approach vs what one would expect with a classic Python approach, thanks @yarikoptic.

> But then taken together - it remains unclear when to raise Exception vs when yield record(s).

I agree, but perhaps that uncertainty is okay and decision should be left up to the developer implementing the specific code (and the subsequent review process) ?